unit/util/expr_initializer.cpp

```diff
@@ -0,0 +1,274 @@
+  auto test = expr_initializer_test_environmentt::make();
+  typet input = bool_typet{};
+  const auto result = nondet_initializer(input, test.loc, test.ns);
+  REQUIRE(result.has_value());
+  const auto expected = side_effect_expr_nondett{bool_typet{}, test.loc};
+  REQUIRE(result.value() == expected);
+}
+
+TEST_CASE("nondet_initializer signed_bv", "[core][util][expr_initializer]")
+{
+  auto test = expr_initializer_test_environmentt::make();
+  typet input = signedbv_typet{8};
+  const auto result = nondet_initializer(input, test.loc, test.ns);
+  REQUIRE(result.has_value());
+  const auto expected = side_effect_expr_nondett{signedbv_typet{8}, test.loc};
+  REQUIRE(result.value() == expected);
+}
+
+TEST_CASE("nondet_initializer c_enum", "[core][util][expr_initializer]")
+{
+  auto test = expr_initializer_test_environmentt::make();
+  const unsignedbv_typet enum_underlying_type{8};
+  const auto enum_type = make_c_enum_type(enum_underlying_type, 3);
+  const auto result = nondet_initializer(enum_type, test.loc, test.ns);
+  REQUIRE(result.has_value());
+  const auto expected = side_effect_expr_nondett{enum_type, test.loc};
+  REQUIRE(result.value() == expected);
+
+  // Repeat with the c_enum_tag_typet instead of the c_enum_typet it points to
+  const symbolt symbol = create_tag_populate_env(enum_type, test);
+  const auto tag_result = nondet_initializer(symbol.type, test.loc, test.ns);
+  REQUIRE(tag_result.has_value());
+  const auto expected_tag = side_effect_expr_nondett{symbol.type, test.loc};
+  REQUIRE(tag_result.value() == expected_tag);
+}
+
+TEST_CASE(
+  "nondet_initializer simple array_type",
+  "[core][util][expr_initializer]")
+{
+  auto test = expr_initializer_test_environmentt::make();
+  typet inner_type = signedbv_typet{8};
+  const std::size_t elem_count = 3;
+  typet array_type =
+    array_typet{inner_type, from_integer(elem_count, signedbv_typet{8})};
+  const auto result = nondet_initializer(array_type, test.loc, test.ns);
+  REQUIRE(result.has_value());
+  std::vector<exprt> array_values{
+    elem_count, side_effect_expr_nondett{signedbv_typet{8}, test.loc}};
+  const auto expected = array_exprt{
+    array_values,
+    array_typet{
+      signedbv_typet{8}, from_integer(elem_count, signedbv_typet{8})}};
+  REQUIRE(result.value() == expected);
+}
+
+TEST_CASE(
+  "nondet_initializer on array_type with nondet size",
+  "[core][util][expr_initializer]")
+{
+  auto test = expr_initializer_test_environmentt::make();
+  typet inner_type = signedbv_typet{8};
+  typet array_type = array_typet{
+    inner_type, side_effect_expr_nondett{signedbv_typet{8}, test.loc}};
+  const auto result = nondet_initializer(array_type, test.loc, test.ns);
+  REQUIRE(result.has_value());
+  const auto expected = side_effect_expr_nondett{
+    array_typet{
+      inner_type, side_effect_expr_nondett{signedbv_typet{8}, test.loc}},
+    test.loc};
+  REQUIRE(result.value() == expected);
+}
+
+TEST_CASE(
+  "nondet_initializer double array_type",
```

Comment:
:pick: "double" here could be confused with the floating point `double` type on first glance. I suggest using the wording "array of arrays".